ai-service/ai_service/app.py

```diff
@@ -1,29 +1,62 @@
 import json
 import logging
+import threading
 from concurrent import futures
+from queue import Queue
+from typing import Generator
 
 import grpc
+from dotenv import load_dotenv
+from langchain.callbacks.base import CallbackManager
+from langchain.callbacks.streaming_stdout import StreamingStdOutCallbackHandler
 from langchain.chat_models import ChatOpenAI
 from langchain.schema import AIMessage, HumanMessage, SystemMessage
 
+import ai_service.gen.teleport.assistant.v1.assistant_pb2_grpc as assistant_grpc
+import ai_service.model as model
 from ai_service.gen.teleport.assistant.v1.assistant_pb2 import (
     CompleteRequest,
     CompletionResponse,
 )
-import ai_service.gen.teleport.assistant.v1.assistant_pb2_grpc as assistant_grpc
-import ai_service.model as model
 
-chat_llm = ChatOpenAI(model_name="gpt-4", temperature=0.5)
+
+class ChainStreamHandler(StreamingStdOutCallbackHandler):
+    def __init__(self, gen: Queue):
+        super().__init__()
+        self.gen = gen
+
+    def on_llm_new_token(self, token: str, **kwargs):
+        self.gen.put(token)
+
 
 DEFAULT_HELLO_MESSAGE = "Hey, I'm Teleport - a powerful tool that can assist you in managing your Teleport cluster via ChatGPT."
 
 
+class AssistantServicer(assistant_grpc.AssistantServiceServicer):
+    def __init__(self) -> None:
+        self.queue = Queue()
+
+        self.chat_llm = ChatOpenAI(
+            model_name="gpt-4",
+            temperature=0.5,
+            streaming=True,
+            callback_manager=CallbackManager([ChainStreamHandler(self.queue)]),
+            verbose=True,  # Some BS, but when verbose is false, then the handler is not being called!!!!!
+        )
+
+    def Complete(self, request, context):
+        logging.debug("Complete called")
+
+        yield from assistant_query(self, request)
+
+
 def assistant_query(
```

Comment:
Is there no way to use eg asyncio here? Would make this code about a *lot* easier to reason about.

Reply:
I think it should be possible.